sql/sys_vars.ic

```diff
@@ -2485,3 +2485,166 @@ public:
   bool global_update(THD *thd, set_var *var);
   uchar *global_value_ptr(THD *thd, const LEX_STRING *base);
 };
+
+
+class Sys_var_vers_asof: public sys_var
+{
+public:
+  Sys_var_vers_asof(
+    const char *name_arg,
+    const char *comment,
+    int flag_args,
+    ptrdiff_t off,
+    size_t size,
+    CMD_LINE getopt,
+    enum charset_enum is_os_charset_arg,
+    const char *def_val,
+    on_check_function on_check_func=0,
+    on_update_function on_update_func=0) :
+  sys_var(
+    &all_sys_vars,
+    name_arg,
+    comment,
+    flag_args,
+    off,
+    getopt.id,
+    getopt.arg_type,
+    SHOW_CHAR,
+    (intptr) def_val,
+    0,
+    VARIABLE_NOT_IN_BINLOG,
+    on_check_func,
+    on_update_func,
+    0)
+  {
+    option.var_type|= GET_STR;
+    if (global_update(def_val))
+    {
+      DBUG_ASSERT(false);
+    }
+  }
+
+  bool do_check(THD *thd, set_var *var)
+  { return false; }
+
+  bool update(String &in, st_vers_current_time &out)
+  {
+    if (in.length() == 3 &&
+      0 == my_strcasecmp(
+        in.charset(),
+        "ALL",
+        in.ptr()))
+    {
+      out.type= FOR_SYSTEM_TIME_ALL;
+    }
+    else if (in.length() == 3 &&
+      0 == my_strcasecmp(
+        in.charset(),
+        "NOW",
+        in.ptr()))
+    {
+      out.type= FOR_SYSTEM_TIME_UNSPECIFIED;
+    }
+    else
+    {
+      MYSQL_TIME_STATUS status;
+      if (str_to_datetime(
+            in.ptr(),
+            in.length(),
+            &out.ltime,
+            flags,
+            &status) ||
+          out.ltime.time_type != MYSQL_TIMESTAMP_DATETIME ||
+          (status.warnings & ~MYSQL_TIME_NOTE_TRUNCATED) != 0)
+      {
+        return true;
+      }
+      out.type= FOR_SYSTEM_TIME_AS_OF;
+    }
+    return false;
+  }
+  bool update(THD *thd, set_var *var, st_vers_current_time &out)
+  {
+    Item *item= var->value;
+
+    switch (item->type())
+    {
+    case Item::DATE_ITEM:
+      {
+        if (item->get_date(&out.ltime, 0))
+          break;
+        out.type= FOR_SYSTEM_TIME_AS_OF;
+        return false;
+      }
+
+    case Item::STRING_ITEM:
+      {
+        String *str= item->val_str();
+        DBUG_ASSERT(str);
+        if (update(*str, out))
+          break;
+        return false;
+      }
+    default:
+      break;
+    }
+    String *str= item->val_str();
+    const char *cstr= str ? str->c_ptr_safe() : "NULL";
+    my_error(ER_WRONG_VALUE_FOR_VAR, MYF(0), var->var->name.str, cstr);
```

Comment:
```sql
set @ts = now(6);
set vers_current_time = @ts;

mysqltest: At line 7: query 'set vers_current_time = @ts' failed: 1231: Variable 'vers_current_time' can't be set to the value of '2017-03-30 16:12:52.634962'

Probably, instead of `item->type()` you should call `item->result_type()`.
```